src/libstore/derivations.cc

```diff
@@ -1097,36 +1097,52 @@ Derivation Derivation::fromJSON(
     const Store & store,
     const nlohmann::json & json)
 {
+    using nlohmann::detail::value_t;
+
     Derivation res;
 
-    res.name = json["name"];
+    ensureType(json, value_t::object);
 
-    {
-        auto & outputsObj = json["outputs"];
-        for (auto & [outputName, output] : outputsObj.items()) {
+    res.name = valueOfTypeAt(json, value_t::string, "name");
+
+    try {
+        auto & outputsObj = valueOfTypeAt(json, value_t::object, "outputs");
+        for (auto & [outputName, _] : outputsObj.items()) {
+            auto & output = valueOfTypeAt(outputsObj, value_t::object, outputName);
             res.outputs.insert_or_assign(
                 outputName,
                 DerivationOutput::fromJSON(store, res.name, outputName, output));
         }
+    } catch (Error & e) {
+        e.addTrace({}, "while reading key 'outputs'");
+        throw;
     }
 
-    {
-        auto & inputsList = json["inputSrcs"];
+    try {
+        auto & inputsList = valueOfTypeAt(json, value_t::array, "inputSrcs");
         for (auto & input : inputsList)
             res.inputSrcs.insert(store.parseStorePath(static_cast<const std::string &>(input)));
+    } catch (Error & e) {
+        e.addTrace({}, "while reading key 'inputSrcs'");
+        throw;
     }
 
-    {
-        auto & inputDrvsObj = json["inputDrvs"];
-        for (auto & [inputDrvPath, inputOutputs] : inputDrvsObj.items())
+    try {
+        auto & inputDrvsObj = valueOfTypeAt(json, value_t::object, "inputDrvs");
```

Comment:
Redundant `try` again because using combo method

Reply:
Fixed here as well by not using the combo method.